src/core/experimental/rcptr.d

```diff
@@ -0,0 +1,208 @@
+// Written in the D programming language.
+/**
+This module provides a shared pointer implementation with memory management
+through reference counting.
+
+License: $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Les De Ridder
+
+Source: $(DRUNTIMESRC core/experimental/rcptr.d)
+*/
+module core.experimental.rcptr;
+
+///
+unittest
+{
+    import core.stdc.stdlib : calloc;
+    auto allocInts = (size_t count) => cast(int*) calloc(count, int.sizeof);
+
+    __rcptr!int a; //default ctor
+    assert(a.get is null);
+
+    auto b = __rcptr!int(allocInts(10)); //ptr ctor
+
+    {
+        auto c = a; //copy ctor
+        assert(c.get == a.get);
+        assert(c.get != b.get);
+
+        c = b; //opAssign
+        assert(c.get != a.get);
+        assert(c.get == b.get);
+    }
+
+    assert(a.count is null);
+    assert(*b.count == 0);
+}
+
+/**
+A reference counted shared pointer that can be used to implement reference
+counted data structures.
+
+The reference count is automatically incremented/decremented on assignment,
+copy (construction), and destruction. When there are no more references to the
+pointer, the reference count is automatically deallocated and the referenced
+pointer is `free`d.
+
+Implementation: The internal implementation of `__rcptr` uses `malloc`/`free`.
+*/
+struct __rcptr(T)
+{
+    alias CounterType = int;
+
+    private T* ptr = null;
+    private shared(CounterType)* count = null;
+
+    /**
+    Creates a new `__rcptr` instance, tracking the provided pointer.
+
+    This implies that the ownership of the pointer is transferred to
+    `__rcptr`.
+
+    Params:
+         ptr = pointer to memory to be managed by `__rcptr`
+    */
+    this(inout(T*) ptr) inout
+    {
+        this.ptr = ptr;
+
+        if (ptr !is null)
+        {
+            import core.memory : pureCalloc;
+
+            // We use `calloc` so we don't have to manually initialise count/addRef
+            count = (() @trusted inout => cast(typeof(count)) pureCalloc(1, CounterType.sizeof))();
+        }
+    }
+
+    /**
+    Get the owned pointer.
+
+    Returns:
+        The pointer owned by this instance of `__rcptr`.
+    */
+    @system
+    inout(T*) get() inout
+    {
+        return ptr;
+    }
+
+    ~this()
+    {
+        delRef();
+    }
+
+    ///
+    void opAssign(ref __rcptr!T rhs)
+    {
+        if (rhs.count == count)
+        {
+            return;
+        }
+
+        delRef();
```

Comment:
You need to `addRef` before `delRef`. It's possible that the `delRef` may destroy `this` before `addRef` down below retains it.

Reply:
We're in `opAssign`, the `addRef` is after changing the counter over.

Reply:
`rhs` might be dead.
You need to:
```d
rhs.addRef();
swap(this, rhs);
rhs.decRef();
```

Reply:
`rhs` is a `ref`, which means the calling stack frame might not have acquired a ref for it.

Reply:
An alternative may be to accept the argument by value...